src/index.js

```diff
@@ -1,56 +1,246 @@
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  init (config, callback) {
+    log('initializing at: %s', this.path)
+    series([
+      (cb) => this.config.set(config, cb),
+      (cb) => this.version.set(repoVersion, cb)
+    ], callback)
+  }
+
+  /**
+   * Open the repo. If the repo is already open no action will be taken.
+   * If the repo is not initialized it will return an error.
+   *
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  open (callback) {
+    if (!this.closed) {
+      return setImmediate(callback)
+    }
+    log('opening at: %s', this.path)
+
+    // check if the repo is already initialized
+    waterfall([
+      (cb) => this._isInitialized(cb),
+      (cb) => lock.lock(this.path, cb),
+      (lck, cb) => {
+        log('aquired repo.lock')
+        this.lockfile = lck
+
+        log('creating flatfs')
+        const FsStore = this.options.fs
+        const s = new FsStore(path.join(this.path, flatfsDirectory), this._fsOptions)
+
+        if (this.options.sharding) {
+          const shard = new core.shard.NextToLast(2)
+          ShardingStore.createOrOpen(s, shard, cb)
+        } else {
+          cb(null, s)
+        }
+      },
+      (flatfs, cb) => {
+        log('Flatfs store opened')
+        this.store = new MountStore([{
+          prefix: new Key(flatfsDirectory),
+          datastore: flatfs
+        }, {
+          prefix: new Key('/'),
+          datastore: new LevelStore(path.join(this.path, levelDirectory), {
+            db: this.options.level
+          })
+        }])
 
-    const blobStores = initializeBlobStores(options.stores)
+        this.blockstore = blockstore(this)
+        this.closed = false
+        cb()
+      }
+    ], (err) => {
+      if (err && this.lockfile) {
+        return this.lockfile.close((err2) => {
+          log('error removing lock', err2)
+          callback(err)
+        })
+      }
+
+      callback(err)
+    })
+  }
 
-    const setup = (name, needs) => {
-      needs = needs || {}
-      const args = [repoPath, blobStores[name]]
-      if (needs.locks) {
-        args.push(this.locks)
+  /**
+   * Check if the repo is already initialized.
+   *
+   * @private
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  _isInitialized (callback) {
+    parallel([
+      (cb) => this.config.exists(cb),
+      (cb) => this.version.check(repoVersion, cb)
+    ], (err, res) => {
+      if (err) {
+        return callback(err)
       }
 
-      if (needs.config) {
-        args.push(this.config)
+      if (!res[0]) {
+        return callback(new Error('repo is not initialized yet'))
       }
+      callback()
+    })
+  }
 
-      return stores[name].setUp.apply(stores[name], args)
+  /**
+   * Close the repo and cleanup.
+   *
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  close (callback) {
+    if (this.closed) {
+      return callback(new Error('repo is already closed'))
     }
```

Comment:
Different behaviour. open will return silently if it is already open, but close will return an error. Or return an error in both or silent in both.